### DISCUSSION TITLE
How to set NODE_ENV at build time

### DISCUSSION THREAD

**Discussion by malimccalla** [vercel/next.js#13410] [Help] [answered]
I have an api endpoint in my next js app - it can be one of three values

- `http://localhost:4000` (development)
- `http://localhost:8080` (test)
- `https://api.example.com` (production)

How can I dynamically change this value dependant on my `NODE_ENV`? For example in a CI environment I need to build my code and have the test api endpoint of `http://localhost:8080`

**Comment by malimccalla** [comment]
It may be worth mentioning that I also need to set a 3rd party public API key (lets say `API_PUBLIC_KEY`) dependant on environment. 

In every environment apart from production it needs to be set to the sandbox api key (including in staging) and the live api key only in production

**Comment by lfades** [comment, marked as answer]
Like this:

```bash
NODE_ENV=production next dev
# or
NODE_ENV=production next start
# or
NODE_ENV=production next build
```

`next start` and `next build` will set `NODE_ENV` to `production` by default, but if there's a custom `NODE_ENV` it won't be replaced.

**Reply by lfades** [reply]
If using Windows (and without WSL) you can do this instead:

```bash
npx cross-env NODE_ENV=production next start
```

**Reply by lfades** [reply]
@malimccalla A note here, `NODE_ENV` should be one of `production` / `development` / `test`, and you should not need to edit it.

If you have to choose a different API url try setting a `API_URL` environment variable or something similar, overall it is not a good idea to change `NODE_ENV` as it may have unexpected side effects

**Reply by lfades** [reply]
Next.js also has built in support for [Environment Variables](https://nextjs.org/docs/basic-features/environment-variables) using `.env` files to make this even easier 💯

**Reply by isaacloft** [reply]
> Like this:
> 
> ```shell
> NODE_ENV=production next dev
> # or
> NODE_ENV=production next start
> # or
> NODE_ENV=production next build
> ```
> 
> `next start` and `next build` will set `NODE_ENV` to `production` by default, but if there's a custom `NODE_ENV` it won't be replaced.

Hi there, thanks for the info above. Im wondering if it's also documented somewhere? I couldnt find similar words in https://nextjs.org/docs/basic-features/environment-variables

**Reply by RIP21** [reply]
Actually, nowadays (next@12x) it's not working. `start` and `build` commands override `NODE_ENV` to be `production` no matter what you do. So inside the bundle after the `build` all the `process.env.NODE_ENV` are getting replaced with `"production"` so all conditionals fail. Maybe `next start` doesn't override it, but it doesn't really matter as all the code has all `NODE_ENV` replaced with a static hardcoded value :)

So `NEXT_ENV` or a similar custom env variable is a way to go.

**Reply by rexwangcc** [reply]
@RIP21 's answer really saved my day! :tada:

**Reply by thescientist13** [reply]
@RIP21 
Is there any way around this that you know of?  Trying to adopt Next within an existing CI / CD system and `NODE_ENV` is already "claimed" at the infrastructure level.

I saw a thread where this could potentially be disabled with webpack, so was going to give that a go. 🤞 
https://github.com/webpack/webpack/issues/7470#issuecomment-394259698

**Reply by RIP21** [reply]
@thescientist13 my answer literally tells what to do :D You have to use a custom env, otherwise, it's always `production`

**Reply by thescientist13** [reply]
Yeah, was just wondering if anyone had found another way around the issue, at least as interim solution.  Sounds like there might not be any escape hatch at all.

**Reply by thescientist13** [reply]
Phew, our team already had something for this in place, so the info I need was available in another environment variable. 😮‍💨

**Reply by bennycode** [reply]
When using `NODE_ENV=development next build`, then I am receiving a lot of building errors like: 

> Error occurred prerendering page "/signin". Read more: https://nextjs.org/docs/messages/prerender-error
> TypeError: Cannot read properties of null (reading 'useContext')
